### DISCUSSION TITLE
How can I display a confirmation message after requesting a new code when using the TOTP strategy?

### DISCUSSION THREAD

**Discussion by andersr** [sergiodxa/remix-auth#245] [Q&A]
I am currently working on integrating the [totp strategy](https://github.com/dev-xo/remix-auth-totp) and had a question re. requesting new codes.  Currently, when following the code in the docs and [the starter example](https://github.com/dev-xo/totp-starter-example), there is no user feedback provided when requesting a new code.  How can I display something like a "new code sent" flash message after a new code is requested? 

Thanks for any tips or insights!
-Anders

**Comment by ersinakinci** [comment]
I'm wondering the exact same thing. Did you find a solution, @andersr?

**Reply by andersr** [reply]
@ersinakinci - my solution was to implement my own auth layer on top of [Remix Sessions](https://remix.run/docs/en/main/utils/sessions) (on top of which remix auth also is built.)  Remix sessions are well-documented and quite straightforward to build on top of.  That may not be the right choice for everyone, but I've been happy with my decision so far.

**Reply by ersinakinci** [reply]
Thanks for sharing!

I'm new to Remix and ended up scratching my head quite a bit. The core thing confusing me was why on earth doesn't Remix Auth offer a generic way to set the session flash. It seems obvious since all the strategies are based on redirecting the user on success/failure. Requesting a new TOTP code also results in a redirect.

The breakthrough I had shortly after asking you my question was realizing that the redirect is going to be handled by the client, meaning that it won't result in a full page refresh and that I don't need to use session flash. I can handle any user notification within React.

I ended up using an `onClick` handler on my login button to modify some React state and display some UI. Hope that helps someone else, wasn't obvious as a newcomer.

**Reply by sergiodxa** [reply]
> The core thing confusing me was why on earth doesn't Remix Auth offer a generic way to set the session flash

Because you have access to the sessionStorage so you can use that

**Reply by ersinakinci** [reply]
@sergiodxa can you give an example of how you would use sessionStorage to accomplish a flash on a successful auth?

If I'm using a cookie session store, then I have to set my cookies after executing `await sessionStorage.commitSession(session)` in the redirect response. But Remix Auth takes over the the entire response, so how would I set the cookies?

**Reply by sergiodxa** [reply]
If you don't pass a successRedirect to the `authenticate` method it will return the user data (what your strategy returned) and you can save it in the session yourself.

The README has an example to decide where to redirect based on the user data https://github.com/sergiodxa/remix-auth?tab=readme-ov-file#custom-redirect-url-based-on-the-user.

You can do something similar, get the user, save it in the sessionKey and add any other flash message you want, you could even return json instead of a redirect.

**Reply by ersinakinci** [reply]
Ah yes, I remember the problem now. The remix-auth strategies that I tried using don't respect your design and make the `successRedirect` option mandatory. Examples:

- [remix-auth-email-link](https://github.com/pbteja1998/remix-auth-email-link/blob/1c08123c3b7ff8a1c23944731ebd45296e0f010d/src/index.ts#L200-L204) ([see issue](https://github.com/pbteja1998/remix-auth-email-link/issues/45))
- [remix-auth-totp](https://github.com/dev-xo/remix-auth-totp/blob/17329075cf20a685d96cce1e87a0dcbcc01e8ede/src/index.ts#L310)

In fairness to the strategy authors, the `successRedirect` contract isn't explicitly spelled out, so it's easy to forget to make it optional.

**Comment by sergiodxa** [comment]
I totally missed this one, since I didn't built the TOTP strategy myself, I think it was better to ask directly in the repo of the strategy.
